core/src/main/java/com/linecorp/armeria/common/util/AsyncLoaderBuilder.java

```diff
@@ -0,0 +1,107 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A builder for creating a new {@link AsyncLoader}.
+ *
+ * <p>Expiration should be set by {@link #expireAfterLoad(Duration)} or {@link #expireIf(Predicate)}.
+ * If expiration is not set, {@link #build()} will throw {@link IllegalStateException}.
+ */
+public final class AsyncLoaderBuilder<T> {
+
+    private final Function<@Nullable T, CompletableFuture<T>> loader;
+    @Nullable
+    private Duration expireAfterLoad;
+    @Nullable
+    private Predicate<@Nullable T> expireIf;
+    @Nullable
+    private Predicate<@Nullable T> refreshIf;
+    @Nullable
+    private BiFunction<Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler;
+
+    AsyncLoaderBuilder(Function<@Nullable T, CompletableFuture<T>> loader) {
+        requireNonNull(loader, "loader");
+        this.loader = loader;
+    }
+
+    /**
+     * Expires loaded value after duration since it was loaded.
+     * New value will be loaded by loader on next {@link AsyncLoader#get()}.
+     */
+    public AsyncLoaderBuilder<T> expireAfterLoad(Duration expireAfterLoad) {
+        requireNonNull(expireAfterLoad, "expireAfterLoad");
+        this.expireAfterLoad = expireAfterLoad;
+        return this;
+    }
+
+    /**
+     * Expires loaded value if predicate matches.
+     * New value will be loaded by loader on next {@link AsyncLoader#get()}.
+     */
+    public AsyncLoaderBuilder<T> expireIf(Predicate<@Nullable T> expireIf) {
+        requireNonNull(expireIf, "expireIf");
+        this.expireIf = expireIf;
+        return this;
+    }
+
+    /**
+     * Refreshes loaded value which is not expired yet asynchronously if predicate matches.
+     * This pre-fetch strategy can remove an additional loading time on a cache miss.
+     *
+     * <p>Note that if 1 refresh is in progress, other refreshes will be bypassed.
+     * Only 1 refresh is executed at the same time.
+     */
+    public AsyncLoaderBuilder<T> refreshIf(Predicate<@Nullable T> refreshIf) {
+        requireNonNull(refreshIf, "refreshIf");
+        this.refreshIf = refreshIf;
+        return this;
+    }
+
+    /**
+     * Handles exception thrown by loader.
+     * If exception handler returns {@code null}, complete {@link AsyncLoader#get()} exceptionally.
+     */
+    public AsyncLoaderBuilder<T> exceptionHandler(BiFunction<
+            Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler) {
```

Comment:
Although, it looks super verbose. 😅
```suggestion
    public AsyncLoaderBuilder<T> exceptionHandler(BiFunction<
            ? super Throwable, ? super @Nullable T, ? extends @Nullable CompletableFuture<T>> exceptionHandler) {
```

Reply:
✅ updated